subprojects/core/src/main/java/org/gradle/internal/classpath/intercept/CallInterceptor.java

```diff
@@ -63,20 +42,9 @@ protected CallInterceptor(InterceptScope... interceptScopes) {
      * @return the value to return to the caller
      * @throws Throwable if necessary to propagate it to the caller
      */
-    public abstract Object doIntercept(Invocation invocation, String consumer) throws Throwable;
-
-    MethodHandle decorateMethodHandle(MethodHandle original, MethodHandles.Lookup caller, int flags) {
-        MethodHandle spreader = original.asSpreader(Object[].class, original.type().parameterCount());
-        MethodHandle decorated = MethodHandles.insertArguments(INTERCEPTOR, 0, this, spreader, flags, caller.lookupClass().getName());
-        return decorated.asCollector(Object[].class, original.type().parameterCount()).asType(original.type());
-    }
+    Object doIntercept(Invocation invocation, String consumer) throws Throwable;
 
-    private Object interceptMethodHandle(MethodHandle original, int flags, String consumer, Object[] args) throws Throwable {
-        boolean isSpread = (flags & IndyInterface.SPREAD_CALL) != 0;
-        return doIntercept(new MethodHandleInvocation(original, args, isSpread), consumer);
-    }
+    MethodHandle decorateMethodHandle(MethodHandle original, MethodHandles.Lookup caller, int flags);
 
-    InterceptScope[] getInterceptScopes() {
-        return interceptScopes;
-    }
+    InterceptScope[] getInterceptScopes();
```

Comment:
❓ Is there a reason to return a mutable array here instead of a list (that internally can be immutable)? We could even return a `Stream<InterceptScope>`; that's immutable by definition, and is easy to compose.

Reply:
@mlopatkin any thoughts?

Reply:
I think we can use list here

Reply:
> @mlopatkin Do you remember the reason?

Probably to reduce overhead when iterating. Given that we're only using it when building the dispatch map, we can indeed return a list/stream here.

Reply:
I figured `Set<InterceptScope>` would make more sense. So I changed to `Set<InterceptScope>`